tests/lifecycle_wait_test.go

```diff
@@ -14,366 +14,546 @@ import (
+//                     It does not wait for the async scripts or
+//                     the other network requests to complete.
+//
+// - networkidle: This lifecycle event is emitted by the browser once:
+//            1. The HTML is loaded;
+//            2. The async scripts have loaded;
+//            3. All other network requests have completed;
+
+func TestLifecycleWaitForNavigation(t *testing.T) {
 	// Test description
 	//
-	// 1. goto /home and wait for the load lifecycle event.
-	// 2. use WaitForLoadState with load to ensure that load
-	//    lifecycle event has already fired.
+	// 1. goto /home and wait for the specified lifecycle event.
+	// 2. click on a link that navigates to a page, and wait on
+	//    the specified lifecycle event.
 	//
-	// Success criteria: We don't wait for all network requests to
-	//                   complete, but we are interested in waiting
-	//                   for all async scripts to have fully loaded
-	//                   (which is when load is fired). We also want
-	//                   to ensure that the load event is stored
-	//                   internally, and we don't block on WaitForLoadState.
+	// Success criteria: The click will perform a navigation away
+	//                   from the current page, it should wait for
+	//                   the specified lifecycle event and the result
+	//                   of the page should match the original nav.
 
 	t.Parallel()
 
-	tb := newTestBrowser(t, withFileServer())
-	p := tb.NewPage(nil)
-	tb.withHandler("/home", func(w http.ResponseWriter, r *http.Request) {
-		http.Redirect(w, r, tb.staticURL("wait_for_nav_lifecycle.html"), http.StatusMovedPermanently)
-	})
-
-	var counter int64
-	var counterMu sync.Mutex
-	tb.withHandler("/ping", func(w http.ResponseWriter, _ *http.Request) {
-		counterMu.Lock()
-		defer counterMu.Unlock()
-
-		time.Sleep(time.Millisecond * 100)
-
-		counter++
-		fmt.Fprintf(w, "pong %d", counter)
-	})
-
-	tb.withHandler("/ping.js", func(w http.ResponseWriter, _ *http.Request) {
-		fmt.Fprintf(w, `
-				var pingJSTextOutput = document.getElementById("pingJSText");
-				pingJSTextOutput.innerText = "ping.js loaded from server";
-			`)
-	})
-
-	waitUntil := common.LifecycleEventLoad
-	assertHome(t, tb, p, waitUntil, func() {
-		result := p.TextContent("#pingRequestText", nil)
-		assert.NotEqualValues(t, "Waiting... pong 10 - for loop complete", result)
-
-		result = p.TextContent("#pingJSText", nil)
-		assert.EqualValues(t, "ping.js loaded from server", result)
-
-		// This shouldn't block and return after calling hasLifecycleEventFired.
-		p.WaitForLoadState(waitUntil.String(), nil)
-	})
+	tests := []struct {
+		name                  string
+		pingSlowness          time.Duration
+		pingJSSlow            bool
+		waitUntil             common.LifecycleEvent
+		pingRequestTextAssert func(result string, pingCount int)
+		pingJSTextAssert      func(result string)
+		assertFunc            func(p api.Page)
```

Comment:
It's good to see that you could find a way to simplify my previous example 👏 Clever!